tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +224,275 @@ static bool HasCanonicalDimensionNumbers(
+    size_t max_operand_sz = 3;  // 2 Inputs + 1 output operand
+    for (auto loop : llvm::enumerate(
+             SmallVector<StringRef>{lhs_loop, rhs_loop, output_loop})) {
+      SmallVector<std::string> loop_dim;
+      size_t pre_elip = loop.value().find(kEllipsis);
+      bool has_elip = pre_elip != std::string::npos;
+      if (!has_elip) pre_elip = loop.value().size();
+      // Adding dim up to ellipsis if it exist
+      for (int pre_ch_ind = 0; pre_ch_ind < pre_elip; pre_ch_ind++)
+        loop_dim.push_back(std::string(1, loop.value()[pre_ch_ind]));
+      if (!has_elip) {
+        all_loop_dim.push_back(std::move(loop_dim));
+        continue;
+      }
+      // Case where Ellipsis presence:
+      size_t post_elip = loop.value().size() - (pre_elip + kEllipsis.size());
+      size_t operand_rank = loop.index() < (max_operand_sz - 1)
+                                ? get_rank(args[loop.index()])
+                                : result_ty.getRank();
+      size_t batch_rank = operand_rank - (pre_elip + post_elip);
+      // Add ("0",...,"N") where N is rank of batch into the loop
+      for (int batch_ind = 0; batch_ind < batch_rank; batch_ind++)
+        loop_dim.push_back(std::to_string(batch_ind));
+      // Add the dimension after ellipsis into the loop
+      for (int post_ch_ind = pre_elip + kEllipsis.size();
+           post_ch_ind < loop.value().size(); post_ch_ind++)
+        loop_dim.push_back(std::string(1, loop.value()[post_ch_ind]));
+      batch_rank_vec.push_back(batch_rank);
+      all_loop_dim.push_back(std::move(loop_dim));
+    }
+    // Check that all ellipsis represent same rank of batch
+    if (batch_rank_vec.size() > 1) {
+      if (!std::equal(batch_rank_vec.begin() + 1, batch_rank_vec.end(),
+                      batch_rank_vec.begin())) {
+        return rewriter.notifyMatchFailure(
+            op, "Invalid Elipsis(ellipsis) within Einsum config!");
+      }
+    }
+
+    // Find all unique indices in the input and output
+    llvm::SmallSetVector<StringRef, 4> input_ind;
+    llvm::SmallSetVector<StringRef, 4> output_ind;
+    int num_of_input_operands = args.size();
+    for (const auto& loops_it : llvm::enumerate(all_loop_dim)) {
+      for (const StringRef& dim : loops_it.value()) {
+        if (loops_it.index() < max_operand_sz - 1) {
+          input_ind.insert(dim);
+        } else {
+          output_ind.insert(dim);
+        }
+      }
+    }
+
+    // Check for contraction/summation indices
+    llvm::SmallSetVector<StringRef, 4> reduction_axe =
+        FindSummationAxes(input_ind, output_ind);
+
+    // Find input/output values and types.
+    auto loc = op.getLoc();
+
+    // Setting up Initial buffer for Output Tensor
+    Value output;
+    auto dyn_sizes = ExtractDynamicEinsumSizes(
+        rewriter, loc, adaptor.lhs(), adaptor.rhs(), all_loop_dim[0],
+        all_loop_dim[1], all_loop_dim[2]);
+    output = GetInitTensor(rewriter, loc, result_ty, dyn_sizes);
```

Comment:
We have to fill zero to output if there are reduction loops.

Reply:
done